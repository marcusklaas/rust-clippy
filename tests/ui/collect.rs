#![warn(possible_shortcircuiting_collect)]

use std::iter::FromIterator;

pub fn div(a: i32, b: &[i32]) -> Result<Vec<i32>, String> {
    let option_vec: Vec<_> = b.into_iter()
        .cloned()
        .map(|i| if i != 0 {
            Ok(a / i)
        } else {
            Err("Division by zero!".to_owned())
        })
        .collect();
    let mut int_vec = Vec::new();
    for opt in option_vec {
        int_vec.push(opt?);
    }
    Ok(int_vec)
}

pub fn generic<T>(a: &[T]) {
    // Make sure that our lint also works for generic functions.
    let _result: Vec<_> = a.iter().map(Some).collect();
}

pub fn generic_collection<T, C: FromIterator<T> + FromIterator<Option<T>>>(elem: T) -> C {
    Some(Some(elem)).into_iter().collect()
}

fn main() {
    // We're collecting into an `Option`. Do not trigger lint.
    let _sup: Option<Vec<_>> = (0..5).map(Some).collect();
}
