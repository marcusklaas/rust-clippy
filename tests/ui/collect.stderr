error: you are creating a collection of `Result`s
 --> $DIR/collect.rs:6:21
  |
6 |     let option_vec: Vec<_> = b.iter()
  |                     ^^^^^^
  |
  = note: `-D clippy::possible-shortcircuiting-collect` implied by `-D warnings`
help: if you are only interested in the case where all values are `Ok`, try
  |
6 |     let option_vec: Result<std::vec::Vec<_>, _> = b.iter()
  |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: you are creating a collection of `Option`s
  --> $DIR/collect.rs:23:18
   |
23 |     let _result: Vec<_> = a.iter().map(Some).collect();
   |                  ^^^^^^
help: if you are only interested in the case where all values are `Some`, try
   |
23 |     let _result: Option<std::vec::Vec<_>> = a.iter().map(Some).collect();
   |                  ^^^^^^^^^^^^^^^^^^^^^^^^

error: you are creating a collection of `Option`s
  --> $DIR/collect.rs:27:34
   |
27 |     Some(Some(elem)).into_iter().collect()
   |                                  ^^^^^^^^^
help: if you are only interested in the case where all values are `Some`, try
   |
27 |     Some(Some(elem)).into_iter().collect::<Option<C>>()
   |                                  ^^^^^^^^^^^^^^^^^^^^^^

error: aborting due to 3 previous errors

